[FEATURE] Document Controllers, Registration, Validation and View (#6571)
TYPO3-Documentation/TYPO3CMS-Reference-CoreApi: Documentation/ExtensionArchitecture/ExtbaseRewrite/Appendix/CommonPitfalls.rst

token, add an :php:`initialize*Action()` method and call
:php:`allowProperties()` (or :php:`allowAllProperties()`) on the relevant
argument's property mapping configuration.

..  seealso::

    `Manually allowing properties on Extbase action arguments <https://docs.typo3.org/permalink/extbase-controller-propertymapping-allowproperties>`_
    — how to write the initializer and which methods to use.

..  _extbase-appendix-pitfalls-template-empty:

Template variable renders empty in a Fluid template
====================================================

**Symptom:** A variable or property path in a Fluid template has no output.
No error is thrown and there is no exception in the log. The surrounding HTML
renders normally; only the value is blank.

**Why:** Fluid silently renders an empty string when property resolution fails.
Resolution is attempted in the following order: the public property, :php:`getX()`,
:php:`hasX()`, :php:`isX()`. If none of these exist or they are
accessible, Fluid gives up without raising an error. Common causes:

*   A typo in the property name — :html:`{conference.titel}` instead of
    :html:`{conference.title}` produces nothing.

> *   A :php:`private` property — Fluid can never access :php:`private`

Reviewer: ```suggestion
*   A :php:`private` property in a model — Extbase cannot hydrate :php:`private`
```

Author: done

    hashing mechanism.

..  _extbase-appendix-pitfalls-property-mapping-denied:

Property mapping denied: form fields not saved without a trusted-properties token
==================================================================================

**Symptom:** A domain object argument arrives in the action with all properties
at their default values even though the form or URL contained data. No error
or validation failure is shown.

> **Why:** When a request does not carry a :php:`__trustedProperties` token,

Reviewer: ```suggestion
**Why:** When a request does not carry a`__trustedProperties` token,
```
Trusted propetries is an http parameter so nothing to do with PHP

Author: done

Extbase denies all properties by default to prevent
`mass assignment <https://owasp.org/www-project-web-security-testing-guide/latest/4-Web_Application_Security_Testing/07-Input_Validation_Testing/20-Testing_for_Mass_Assignment>`_
attacks. The token is generated automatically by :html:`<f:form>` and covers
exactly the fields rendered in the form. Requests that bypass this — URL
parameters, custom forms without the token, or JSON payloads — carry no token,
so no property is allowed unless the controller explicitly permits it.

> **What to do:** If the request will never carry a :php:`__trustedProperties`

Reviewer: ```suggestion
**What to do:** If the request will never carry a `__trustedProperties`
```

Author: done

token, add an :php:`initialize*Action()` method and call
:php:`allowProperties()` (or :php:`allowAllProperties()`) on the relevant
argument's property mapping configuration.

..  seealso::

    `Manually allowing properties on Extbase action arguments <https://docs.typo3.org/permalink/extbase-controller-propertymapping-allowproperties>`_
    — how to write the initializer and which methods to use.

..  _extbase-appendix-pitfalls-template-empty:

Template variable renders empty in a Fluid template
====================================================

**Symptom:** A variable or property path in a Fluid template outputs nothing.
No error is thrown and no exception appears in the log. The surrounding HTML
renders normally; only the value is blank.

**Why:** Fluid silently renders an empty string when property resolution fails.
Resolution is attempted in this order: a public property, then :php:`getX()`,
then :php:`hasX()`, then :php:`isX()`. If none of these exist or are

> accessible, Fluid gives up without raising an error. Common causes:

Reviewer 2: What about array keys, if the variable is an array instead of object, maybe mention that too?

Author: done

*   A typo in the property name — :html:`{conference.titel}` instead of
    :html:`{conference.title}` silently produces nothing.

> *   A :php:`private` property — Fluid can never access :php:`private`

Reviewer: Are you sure about this? I think private properties are only a problem when you are working with Extbase models during hydration. when working with DTOs that you initialize yourself or with values that are transient there appears to be no such restriction?

Reviewer 2: If the getter is a public method, that should work...

Author: as much as it pains me, you are both right. Extbase = red pill against the pain, yellow pill against the red one, and blue pill to stay concious. There seems to be no way to state a fact without at least three exceptions.
Better now?

    properties, even if a getter exists on the same class.
*   A missing getter — a :php:`protected` property without a corresponding
    :php:`getX()` method is invisible to Fluid.

> *   The variable was not assigned in the controller — :php:`assign()` was not

Reviewer 3: Maybe a note about missing TCA property?

Author: done

>     called, or was called under a different name.

Reviewer 3: Missing transfer of variable from template/partial to partial

Author: good catch, done

**What to do:** Check the exact property name and visibility. Add a

> :php:`getX()` method if one is missing. In the controller, verify that

Reviewer: ```suggestion
public :php:`getX()` method if one is missing. In the controller, verify that
```

Author: done

:php:`$this->view->assign('name', $value)` is called with the correct variable
name. In the template itself, use
`f:debug <https://docs.typo3.org/permalink/t3viewhelper:typo3-fluid-debug>`_
to dump the value at the point of use:

..  code-block:: html
    :caption: EXT:my_extension/Resources/Private/Templates/Conference/Show.fluid.html

>     <f:debug>{conference}</f:debug>

Reviewer 3: Above you wrote about miss-spelling getTitle, then you write `name` as example and here you write `conference`. Would be nice to streamline the variable name a bit.

Author: done

This renders a formatted dump of the variable including its type and all
accessible properties.

..  _extbase-appendix-pitfalls-fdebug-no-output:

f:debug produces no output
==========================

**Symptom:** :html:`<f:debug>{variable}</f:debug>` is in the template but
nothing appears on the page.

> **Why:** Two independent causes exist and both are common:

Reviewer 2: Another reason: JSON views or inline javascript rendering, and f:debug by default is placed on DOM top.

Maybe reference the `inline="1"` possibilities here, which ensures the debug output occurs exactly at the place of usage and not being prepended to the DOM top.

Author: I did not even know that myself. Thank you, added.

..  _extbase-appendix-pitfalls-template-not-found:

Template file not found, or wrong template rendered
====================================================

**Symptom:** Extbase throws a "Could not find template file" exception, or
renders a template from a different path than expected — for example, a
customised template is ignored and the original one is used instead.

**Why:** Template resolution is based on a numerically keyed path array
searched from the highest key downward. Several things can go wrong:

*   **Wrong key order:** a customisation registered at key :typoscript:`5` is
    ignored in favour of the original at key :typoscript:`10`, because
    :typoscript:`10` is higher and wins.
*   **Case mismatch on Linux:** :file:`List.fluid.html` and
    :file:`list.fluid.html` are different files. The convention requires an
    uppercase first letter for both the controller subdirectory and the action
    file name.
*   **Controller subdirectory name mismatch:** the subdirectory must match the
    controller class name without the :php:`Controller` suffix —
    :php:`ConferenceController` requires :file:`Conference/`, not
    :file:`Conferences/` or :file:`conference/`.

> *   **Extension load order:** when overriding templates from another extension,

Reviewer 3: It's related to the TypoScript loading order, or the Site Set loading order. You need the extension loading order, if you extend the domain models with DI aliases or EXT:extender. Yes, it would be good to respect the loading order, but it has nothing to to with template loading order.

Author: reworked
